src/client/Microsoft.Identity.Client/Region/RegionDiscoveryProvider.cs

```diff
@@ -75,47 +81,52 @@ private async Task<string> GetRegionAsync(RequestContext requestContext)
                 };
-                
-                HttpResponse response = await _httpManager.SendGetAsync(BuildImdsUri(DefaultApiVersion), headers, logger).ConfigureAwait(false);
+
+                HttpResponse response = await _httpManager.SendGetAsync(BuildImdsUri(DefaultApiVersion), headers, logger, retry: false, GetCancellationToken(requestContext.UserCancellationToken)).ConfigureAwait(false);
 
                 // A bad request occurs when the version in the IMDS call is no longer supported.
                 if (response.StatusCode == HttpStatusCode.BadRequest)
                 {
-                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers).ConfigureAwait(false); // Get the latest version
-                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger).ConfigureAwait(false); // Call again with updated version
+                    string apiVersion = await GetImdsUriApiVersionAsync(logger, headers, requestContext.UserCancellationToken).ConfigureAwait(false); // Get the latest version
+                    response = await _httpManager.SendGetAsync(BuildImdsUri(apiVersion), headers, logger, retry: false, GetCancellationToken(requestContext.UserCancellationToken)).ConfigureAwait(false); // Call again with updated version
                 }
 
                 if (response.StatusCode == HttpStatusCode.OK && !response.Body.IsNullOrEmpty())
                 {
-                    LocalImdsResponse localImdsResponse = JsonHelper.DeserializeFromJson<LocalImdsResponse>(response.Body);
-
-                    if (localImdsResponse != null && !localImdsResponse.location.IsNullOrEmpty())
-                    {
-                        logger.Info($"[Region discovery] Call to local IMDS returned region: {localImdsResponse.location}");
-                        LogTelemetryData(localImdsResponse.location, RegionSource.Imds, requestContext);
-
-                        return localImdsResponse.location;
-                    }
+                    return response.Body;
                 }
                     
                 logger.Info($"[Region discovery] Call to local IMDS failed with status code: {response.StatusCode} or an empty response.");
 
                 throw MsalServiceExceptionFactory.FromImdsResponse(
-                MsalError.RegionDiscoveryFailed,
-                MsalErrorMessage.RegionDiscoveryFailed,
-                response);
-            }
-            catch (MsalServiceException)
+                    MsalError.RegionDiscoveryFailed,
+                    MsalErrorMessage.RegionDiscoveryFailed,
+                    response);
+            }  
+            catch (MsalServiceException e)
             {
-                throw;
+                if (MsalError.RequestTimeout.Equals(e.ErrorCode))
+                {
+                    throw new MsalServiceException("region_discovery_failed", "app cancelled or timeout expired", e);
```

Comment:
These should be `MsalError` constants?

Reply:
yes. and MsalErrorMessage. nice catch

---
In reply to: [541345262](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/2283#discussion_r541345262) [](ancestors = 541345262)

Reply:
Good point. Updated.